### DISCUSSION TITLE
Serial connection to Teensy disconnects

### DISCUSSION THREAD

**Discussion by mgraupe** [bonsai-rx/bonsai#1276] [Q&A]
Hello, 

I have programmed a Teensy to perform actions in response to different serial strings (e.g. "1,90" or "1,-90"). Based on key presses, I would like to send (using SerialStringWrite) different strings to the Teensy. I would very much appreciate help with two issues : 

1. I can perform one key press and the action is executed correctly. However, the computer plays a chime after some time (~10 s) and when I press the key again, Bonsai shows an Runtime Error message "The port is closed.". How can I keep the serial connection alive for multiple communications and executions? 

2. How can I have the SerialStringWrite function subscribe to multiple string inputs, based on different key presses. It seem as if MultcastSubject is only for publishing integers.? 

![image](https://user-images.githubusercontent.com/8803795/224062306-b04b09f0-7603-4a65-8bc6-62c3ae9ea19a.png)

Thank you very much in advance. 

Best,
Michael

**Comment by mgraupe** [comment]
And here is my workflow. Sorry, just had to figure out that I needed to zip it first. 
[teensyWrite.zip](https://github.com/bonsai-rx/bonsai/files/10932905/teensyWrite.zip)

**Comment by bruno-f-cruz** [comment]
Hi.
Regarding 1, I don't think that's a problem with Bonsai... It sounds like something weird is going on with your microcontroller, or USB line. Especially if the device is dropping its connection. 

Regarding 2 here s the solution I think:
You might want to change the publish to a behavior subject, in case you send a sequence termination.
[teensyWrite.zip](https://github.com/bonsai-rx/bonsai/files/10933602/teensyWrite.zip)

Cheers,
Bruno

**Comment by mgraupe** [comment]
Dear Bruno, 

Thank you very much for you prompt help. It very much appreciate it. 

1. I can send multiple instructions to the Teensy when using the "Serial Monitor" tool of the Arduino IDE software. So I know that it is possible in general with the system. Could it be that the serial connection parameters are not correct and that the connection is closed after a timeout? 

![Screenshot 2023-03-10 100317](https://user-images.githubusercontent.com/8803795/224272319-f6357ee6-88b0-4c21-b765-3dd4d6bf4ff3.png)

2.  Indeed, publishing it with a behavior subject works but can one instance of "SubscribeSubject" access multiple behavior subjects with the same name? I get an error when I create behavior subjects with the same name. 

Cheers,
Michael

**Reply by bruno-f-cruz** [reply]
1. I honestly don't know what serial parameters the Teensy expects. However, if you can send and receive a single message successfully, the parameters are probably correct. Additionally, even if you have some sort of timeout in Bonsai for the serial connection, the language would just dispose of the connection and not "unplug" the device (which is what sounds like it is happening in your case when you hear the "chime"). If I had to guess, some instruction you are sending from bonsai to teensy is getting it to reset. Maybe some short-circuit is going on when you send a specific instruction? Not sure what hardware you have connected to it...
2. You should not change the multicast, you should change the source PublishSubject node. You can only have one source with a given name. In your example, you did the opposite. Check the documentation for an overview of [subjects](https://bonsai-rx.org/docs/articles/subjects.html)

best,
Bruno

**Reply by mgraupe** [reply]
Dear Bruno, 

Thanks. 
1. Your comment made me remove the "NewLine" setting of SerialStringWrite. It works now. I suppose that the newline "\r\n" made the Teensy reset. 
2. Thanks for the hint, I got it working now with MulticastSubject and BehaviorSubject linking with each other. 

Exciting! And thank you again for the quick help. 

Cheers,
Michael
